mlir-sycl/doc/design/SYCLGridRepresentation.md

```diff
@@ -0,0 +1,244 @@
+
+### `SYCLMethodOpInterfaceImpl`
+
+To implement this interface consistently and avoid code repetition, the TableGen
+`SYCLMethodOpInterfaceImpl` class is provided. Every SYCL dialect operation
+implementing the aforementioned interface must be derived from this class.
+
+### Registering operations implementing `SYCLMethodOpInterface`
+
+The `SYCLDialect::addOperations()` function is defined, overriding
+`Dialect::addOperations()`. This function calls the base implementation, but
+also **registers** operations as `SYCLMethodOpInterface` instances. In order to
+do so, a `MethodRegistry` member variable is added to the dialect. This holds a
+`(TypeID, FunctionName) -> OperationName` table so that operations can be
+registered as "methods" calling `MethodRegistry::registerMethod(mlir::TypeID,
+llvm::StringRef, llvm::StringRef)`, where the first two parameters correspond to
+the keys of the map (`getTypeID()` and `getMethodNames()`) and the last value,
+to the value of the entry (the name of the operation).
+
+This process will require no action from the users of the dialect. As this
+mechanism is implemented in the `SYCLDialect::addOperations()` function,
+defining a new function implementing `SYCLMethodOpInterface` in the TableGen
+file also generates call to register it with no further action.
+
+### Looking up operations implementing `SYCLMethodOpInterface`
+
+In the codegen side, right before generating a `sycl.call` operation, we check
+whether there is an operation implementing `SYCLMethodOpInterface` homologous to
+that function call. In order to do that, `SYCLDialect::findMethod(mlir::TypeID,
+llvm::StringRef)` can be used, passing the ID of the SYCL type the function is a
+member of and the function name. After looking up the name in its registry,
+`llvm::None` is returned if no operation can be used to replace the `sycl.call`;
+otherwise, the name of the operation is returned, which can be used to create a
+different operation instead.
+
+Note that, in case the function is a member of a base type, the `sycl.cast`
+operation used to cast to the base class is abstracted beforehand.
+
+### Canonicalization
+
+To help the analysis, canonicalization rules are introduced in such a way that
+the total number of operations is minimized, i.e., we reduce the number of
+operations by replacing combinations of operations by equivalent
+single-operations, e.g., the following code:
+
+```MLIR
+%0 = sycl.nd_item.get_global_id(%nd) {FunctionArgTypes = {!llvm.ptr<struct<...>, 4>}, FunctionName = @get_global_id, TypeName = @nd_item} : (!sycl_nd_item_1_) -> !sycl_id_1_
+%1 = sycl.id.get(%0, %idx) {FunctionArgTypes = {memref<?x!sycl_id_1_, 4>, i32}, FunctionName = @get, TypeName = @id} : (!sycl_id_1_, i32) -> i64
+```
+
+would be simplified to:
+
+```MLIR
+%1 = sycl.nd_item.get_global_id(%nd, %idx) {FunctionArgTypes = {!llvm.ptr<struct<...>, 4>, i32}, FunctionName = @get_global_id, TypeName = @nd_item} : (!sycl_nd_item_1_, i32) -> i64
```

Comment:
is it possible that this new call is not part of the register?

Reply:
Yes, I missed that explanation. In that case, operations should be "decanonicalized", as I have just added.

Reply:
Note that the canonicalization process we apply here does not have performance as a direct goal, but it tries to improve the analysis, so we do not fear decanonicalization here (honestly, I don't think performance will be affected by either canonicalization/decanonicalization).